assets/js/modules/analytics-4/components/widgets/TopTrafficSourceDrivingPurchasesWidget.js

```diff
@@ -0,0 +1,229 @@
+		metrics: [
+			{
+				name: 'ecommercePurchases',
+			},
+		],
+	};
+
+	const trafficSourceReportOptions = {
+		...dates,
+		dimensions: [ 'sessionDefaultChannelGroup' ],
+		metrics: [
+			{
+				name: 'ecommercePurchases',
+			},
+		],
+		limit: 1,
+		orderBy: 'ecommercePurchases',
+	};
+
+	const totalPurchasesReport = useInViewSelect(
+		( select ) =>
+			hasDetectedEvent
+				? select( MODULES_ANALYTICS_4 ).getReport(
+						totalPurchasesReportOptions
+				  )
+				: undefined,
+		[ hasDetectedEvent, totalPurchasesReportOptions ]
+	);
+	// ecommercePurchases metric will always be assigned a value, unlike most
+	// other metrics, where rows would be empty if there is no data, ecommercePurchases
+	// will assign value `0` for each date range if there is no event data. So we need
+	// to verify that value is non-zero for one of the rows, to allow report request for
+	// traffic source. Otherwise zero data will be incorrectly displayed, showing zero for
+	// purchases and comparison, but still include the traffic source.
+	const hasPurchases = totalPurchasesReport?.rows
+		? totalPurchasesReport?.rows?.[ 0 ]?.metricValues?.[ 0 ]?.value > 0 ||
+		  totalPurchasesReport?.rows?.[ 1 ]?.metricValues?.[ 0 ]?.value > 0
+		: false;
+
+	const trafficSourceReport = useInViewSelect(
+		( select ) =>
+			hasDetectedEvent && hasPurchases
+				? select( MODULES_ANALYTICS_4 ).getReport(
+						trafficSourceReportOptions
+				  )
+				: undefined,
+		[ hasDetectedEvent, hasPurchases, trafficSourceReportOptions ]
+	);
+
+	const error = useSelect( ( select ) => {
+		const trafficSourceReportErrors = select(
+			MODULES_ANALYTICS_4
+		).getErrorForSelector( 'getReport', [ trafficSourceReportOptions ] );
+
+		const totalPurchasesReportErrors = select(
+			MODULES_ANALYTICS_4
+		).getErrorForSelector( 'getReport', [ totalPurchasesReportOptions ] );
+
+		if ( trafficSourceReportErrors && totalPurchasesReportErrors ) {
+			return [ trafficSourceReportErrors, totalPurchasesReportErrors ];
+		}
+
+		return (
+			trafficSourceReportErrors || totalPurchasesReportErrors || undefined
+		);
+	} );
+
+	const loading = useSelect( ( select ) =>
+		hasDetectedEvent
+			? ! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
+					'getReport',
+					[ totalPurchasesReportOptions ]
+			  ) ||
+			  ( ! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
+					'getReport',
+					[ trafficSourceReportOptions ]
+			  ) &&
+					hasPurchases )
+			: undefined
+	);
+
+	const makeFilter = ( dateRange, dimensionIndex ) => ( row ) =>
+		get( row, `dimensionValues.${ dimensionIndex }.value` ) === dateRange;
+
+	// Prevents running a filter on `report.rows` which could be undefined.
+	const { rows: totalPurchasesReportRows = [] } = totalPurchasesReport || {};
+	const { rows: trafficSourceReportRows = [] } = trafficSourceReport || {};
+
+	const topTrafficSource =
+		trafficSourceReportRows.filter( makeFilter( 'date_range_0', 1 ) )[ 0 ]
+			?.dimensionValues?.[ 0 ].value || '-';
```

Comment:
Similar to the add_to_card widget, we need to display a better UI when there is no data for this widget. We can probably show zero data state in this case.

Reply:
I pinged Jimmy in the group, I can't make changes specific to these ACR widgets, as they incorporate existing layout that is used by few initial key metrics, and they need to be consistent across all widgets. I am checking if this is part of initial design, etc, then I will open a follow up issue to suggest update to this layout, and ping Sigal about. But in the meantime I am collecting more info, and we do not need to hold these issues off for this, as they are behind feature flag and the current zero data state layout is already in use for other core metrics. So this can be improved in a new issue that will target all metrics with this layout

Reply:
Thanks, @zutigrm.